xcube_stac/stac.py

```diff
@@ -63,3 +173,107 @@ def open_data(self, data_id: str, **open_params) -> xr.Dataset:
+        """
+
+        if (
+            pystac_object.extra_fields["type"] == "Collection" and
+            pystac_object.id not in open_params.get("collections", [pystac_object.id])
+        ):
+            pass
+        else:
+            if recursive:
+                if any(True for _ in pystac_object.get_children()):
+                    iterators = (self._get_items_nonsearchable_catalog(
+                        child,
+                        recursive=True,
+                        **open_params
+                    ) for child in pystac_object.get_children())
+                    yield from chain(*iterators)
+                else:
+                    iterator = self._get_items_nonsearchable_catalog(
+                        pystac_object,
+                        recursive=False,
+                        **open_params
+                    )
+                    yield from iterator
+            else:
+                for item in pystac_object.get_items():
+                    # test if item's bbox intersects with the desired bbox
+                    if "bbox" in open_params:
+                        if not self._assert_bbox_intersect(item, **open_params):
+                            continue
+                    # test if item fit to desired time range
+                    if "time_range" in open_params:
+                        if not self._assert_datetime(item, **open_params):
+                            continue
+                    # iterate through assets of item
+                    yield item
+
+    def _assert_datetime(self, item: Item, **open_params) -> bool:
+        """Assert if the datetime or datetime range of an item fits to the
+        'time_range' given by *open_params*.
+
+        Args:
+            item: item/feature
+
+        Returns:
+            True, if the datetime of an item is within the 'time_range',
+            otherwise False. True, if there is any overlap between the
+            'time_range' and the datetime range of an item.
+
+        """
+        dt_start = datetime.fromisoformat(
+            open_params["time_range"][0]
+        ).replace(tzinfo=timezone.utc)
+        dt_end = datetime.fromisoformat(
+            open_params["time_range"][1]
+        ).replace(tzinfo=timezone.utc)
+        if item.properties["datetime"] == "null":
+            dt_start_data = datetime.fromisoformat(
+                item.properties["start_datetime"]
+            )
+            dt_end_data = datetime.fromisoformat(
+                item.properties["end_datetime"]
+            )
+            if dt_end < dt_start_data or dt_start > dt_end_data:
+                return False
+            else:
+                return True
+        else:
+            dt_data = datetime.fromisoformat(item.properties["datetime"])
+            if dt_end < dt_data or dt_start > dt_data:
+                return False
+            else:
+                return True
+
+    def _assert_bbox_intersect(self, item: Item, **open_params) -> bool:
```

Comment:
As above, "assert" is inappropriate -- consider something like `do_bboxes_intersect` instead.

Reply:
@konstntokas I elaborated a little in [my comment](https://github.com/xcube-dev/xcube-stac/pull/10#discussion_r1598677834) on `_assert_datetime` above:

> "Assert" usually means something like "evaluate this condition which should always be true, and throw an exception if it's not". Here you're just returning the result rather than acting on it.

This applies equally to both `_assert_datetime` and `_assert_bbox_intersect`: they return boolean values rather than deliberately throwing exceptions if the checked condition is false, so they are not "asserting", they are just evaluating and returning.

By the way, I accidentally omitted the leading underscore in the alternative method names that I suggested -- they should have been `_do_bboxes_intersect` and `_is_datetime_in_range` to indicate that they're not part of the external API.